### PR TITLE
tetragon: init at 0.11.0

### PR DESCRIPTION
**Pull request by gngram**
## Description of changes

Tetragon: real-time, eBPF-based Security Observability and Runtime Enforcement tool.
See:
https://github.com/cilium/tetragon

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/te/tetragon/package.nix [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  };
+
+  buildInputs = [
+    clang_16
+    go
+    llvm_16
+    pkg-config
+  ];
+
+  buildPhase = ''
```
Comment (raspher) [inline pkgs/by-name/te/tetragon/package.nix:29]: runHook preBuild

File: pkgs/by-name/te/tetragon/package.nix [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    export HOME=$TMP
+    export LOCAL_CLANG=1
+    export LOCAL_CLANG_FORMAT=1
+    make tetragon
+    make tetragon-operator
+    make tetra
+    NIX_CFLAGS_COMPILE="-fno-stack-protector -Qunused-arguments" make tetragon-bpf
+  '';
```
Comment (raspher) [inline pkgs/by-name/te/tetragon/package.nix:37]: runHook postBuild

File: pkgs/by-name/te/tetragon/package.nix [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  postPatch = ''
+    substituteInPlace bpf/Makefile --replace '/bin/bash' '${bash}/bin/bash'
+  '';
+
+  installPhase = ''
```
Comment (raspher) [inline pkgs/by-name/te/tetragon/package.nix:43]: runHook preInstall

File: pkgs/by-name/te/tetragon/package.nix [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    mkdir -p $out/lib
+    mkdir -p $out/lib/tetragon
+    sed -i "s+/usr/local/+$out/+g" install/linux-tarball/usr/local/lib/tetragon/tetragon.conf.d/bpf-lib
+    cp -n -r install/linux-tarball/usr/local/lib/tetragon/tetragon.conf.d/ $out/lib/tetragon/
+    cp -n -r ./bpf/objs $out/lib/tetragon/bpf
+    mkdir -p $out/lib/tetragon/tetragon.tp.d/
+    install -m755 -D ./tetra $out/bin/tetra
+    install -m755 -D ./tetragon $out/bin/tetragon
+  '';
```
Comment (raspher) [inline pkgs/by-name/te/tetragon/package.nix:52]: runHook postInstall

**Comment by raspher** [comment]
please squash your 2 last commits as specified in naming conventions

**Comment by gangaram-tii** [comment]
> please squash your 2 last commits as specified in naming conventions

Done, please have a look. Thanks.

**Comment by raspher** [comment]
Package builded on x86_64 with `nixpkgs-review pr X`, executables are present, but i've problem running `tetragon`.
```
time="2024-02-04T13:06:32+01:00" level=fatal msg="Failed to start tetragon" error="tetragon, aborting could not find BPF programs: sensor program \"bpf_exit.o\" can not be found"
```

[full log](https://gist.github.com/raspher/7a72c5fb643fe408618159a885370de5)

**Comment by jtbx** [comment]
Same thing's happening on my side too

**Comment by gangaram-tii** [comment]
I used following command to run tetragon as suggested by  
https://tetragon.io/docs/contribution-guide/development-setup/#run-tetragon

`sudo tetragon --bpf-lib /nix/store/had9684wa3k5j23vfypjsjap5xzb0r2r-tetragon-0.11.0/lib/tetragon/bpf/`

Now I amended default library path in the Tetragon source code.

**Comment by raspher** [comment]
Result of `nixpkgs-review pr 286147` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tetragon</li>
  </ul>
</details>

**Comment by gangaram-tii** [comment]
This is what the usage of the application. Please review and merge it.

File: pkgs/by-name/te/tetragon/package.nix [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+  buildPhase = ''
+    runHook preBuild
+    export HOME=$TMP
+    export LOCAL_CLANG=1
+    export LOCAL_CLANG_FORMAT=1
+    make tetragon
+    make tetragon-operator
+    make tetra
+    NIX_CFLAGS_COMPILE="-fno-stack-protector -Qunused-arguments" make tetragon-bpf
```
Comment (drupol) [inline pkgs/by-name/te/tetragon/package.nix:37]: I think this should be at the root of the attribute set, as such:

`env.NIX_CFLAGS_COMPILE = "-fno-stack-protector -Qunused-arguments";`

File: pkgs/by-name/te/tetragon/package.nix [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    runHook postBuild
+  '';
+
+  postPatch = ''
+    substituteInPlace bpf/Makefile --replace '/bin/bash' '${bash}/bin/bash'
```
Comment (drupol) [inline pkgs/by-name/te/tetragon/package.nix:42]: Have you tried this?

```suggestion
    substituteInPlace bpf/Makefile --replace '/bin/bash' '${lib.getExe bash}'
```

File: pkgs/by-name/te/tetragon/package.nix [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    substituteInPlace pkg/defaults/defaults.go --replace '/var/lib/tetragon/' $out/lib/tetragon/bpf/
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/lib/tetragon
```
Comment (drupol) [inline pkgs/by-name/te/tetragon/package.nix:48]:
```suggestion
    mkdir -p $out/lib/tetragon $out/lib/tetragon/tetragon.tp.d/
```

File: pkgs/by-name/te/tetragon/package.nix [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    sed -i "s+/usr/local/+$out/+g" install/linux-tarball/usr/local/lib/tetragon/tetragon.conf.d/bpf-lib
+    cp -n -r install/linux-tarball/usr/local/lib/tetragon/tetragon.conf.d/ $out/lib/tetragon/
+    cp -n -r ./bpf/objs $out/lib/tetragon/bpf
+    mkdir -p $out/lib/tetragon/tetragon.tp.d/
+    install -m755 -D ./tetra $out/bin/tetra
+    install -m755 -D ./tetragon $out/bin/tetragon
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
Comment (drupol) [inline pkgs/by-name/te/tetragon/package.nix:58]: Can you please sort meta attributes alphabetically?

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I've left some comments, let me know if this is clear enough.

File: pkgs/by-name/te/tetragon/package.nix (hunk 5 above)
Comment (gangaram-tii) [reply pkgs/by-name/te/tetragon/package.nix]: Done.

File: pkgs/by-name/te/tetragon/package.nix (hunk 6 above)
Comment (gangaram-tii) [reply pkgs/by-name/te/tetragon/package.nix]: Done.

File: pkgs/by-name/te/tetragon/package.nix (hunk 7 above)
Comment (gangaram-tii) [reply pkgs/by-name/te/tetragon/package.nix]: Done.

File: pkgs/by-name/te/tetragon/package.nix (hunk 8 above)
Comment (gangaram-tii) [reply pkgs/by-name/te/tetragon/package.nix:59]: Done!

**Review by drupol — APPROVED** [review]

File: pkgs/by-name/te/tetragon/package.nix [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    description = "Real-time, eBPF-based Security Observability and Runtime Enforcement tool";
+    homepage = "https://github.com/cilium/tetragon";
+    license = {
+      url = "https://github.com/cilium/tetragon/blob/main/LICENSE";
+    };
```
Comment (drupol) [inline pkgs/by-name/te/tetragon/package.nix:64]: How about using the existing licence? https://github.com/NixOS/nixpkgs/blob/919cb864b1d7a4bee3d88c801f71daede4be5aec/lib/licenses.nix#L127

Comment (gangaram-tii) [reply pkgs/by-name/te/tetragon/package.nix]: Yes, nice. In fact I tried to find out licenses.apache* but could not find. Amended the commit with existing license.

**Pull request merged**